lib/tpm/tpm.go

```diff
@@ -0,0 +1,216 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tpm
+
+import (
+	"context"
+	"crypto"
+	"crypto/sha256"
+	"crypto/x509"
+	"fmt"
+	"log/slog"
+	"math/big"
+	"strings"
+
+	"github.com/google/go-attestation/attest"
+	"github.com/gravitational/trace"
+	"go.opentelemetry.io/otel"
+)
+
+var tracer = otel.Tracer("github.com/gravitational/teleport/lib/tpm")
+
+// serialString converts a serial number into a readable colon-delimited hex
+// string thats user-readable e.g ab:ab:ab:ff:ff:ff
+func serialString(serial *big.Int) string {
+	hex := serial.Text(16)
+	if len(hex)%2 == 1 {
+		hex = "0" + hex
+	}
+
+	out := strings.Builder{}
+	for i := 0; i < len(hex); i += 2 {
+		if i != 0 {
+			out.WriteString(":")
+		}
+		out.WriteString(hex[i : i+2])
+	}
+	return out.String()
+}
+
+// hashEKPub hashes the public part of an EK key. The key is first marshaled
+// in PKIX format and then hashed with SHA256.
+func hashEKPub(key crypto.PublicKey) (string, error) {
+	marshaled, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	hashed := sha256.Sum256(marshaled)
+	return fmt.Sprintf("%x", hashed), nil
+}
+
+// QueryRes is the result of the TPM query performed by Query.
+type QueryRes struct {
+	// EKPub is the PKIX marshaled public part of the EK.
+	EKPub []byte
+	// EKPubHash is the SHA256 hash of the PKIX marshaled EKPub.
+	EKPubHash string
+	// EKCertPresent is true if an EK cert is present in the TPM.
+	EKCertPresent bool
+	// EKCert is the ASN.1 DER encoded EK cert.
+	EKCert []byte
+	// EKCertSerial is the serial number of the EK cert represented as a colon
+	// delimited hex string.
+	EKCertSerial string
```

Comment:
I mostly went with these because it matches the format that the user will specify them in the JoinToken allow rules and also the format I want to print them when folks use the `tpm identify` commands - I could switch to []byte/*big.Int and perform these conversions in the RPC logic and in the CLI command logic but it doesn't seem to make anything clearer for me.

Reply:
My reasoning is that string formats often change, or we need the same thing represented in various ways, then off we go parsing hex to bytes to turn it into something else. "Raw" formats obligatorily push that noise somewhere else, but are more flexible.

That said I'm sure I've done it both ways, so this is more of a curiosity than a request for change.

Reply:
I think if this was in a Protobuf spec, I'd be more inclined to agree with that sentiment - as a change in future would require additional compatibility work. Since this is purely on the Go side and restricted to the client itself, I think it's pretty trivial for us to change the format here as necessary.